core/tls-websocket-server/src/connection.rs

```diff
@@ -18,96 +18,271 @@
 #[cfg(all(not(feature = "std"), feature = "sgx"))]
 use crate::sgx_reexport_prelude::*;
 
-use crate::{WebSocketConnection, WebSocketError, WebSocketResult};
+use crate::{error::WebSocketError, WebSocketConnection, WebSocketMessageHandler, WebSocketResult};
 use log::*;
-use rustls::ServerSession;
+use mio::{event::Event, net::TcpStream, Poll, Ready, Token};
+use rustls::{ServerSession, Session};
 use std::{
 	format,
-	net::TcpStream,
 	string::{String, ToString},
+	sync::Arc,
 };
 use tungstenite::{accept, Message, WebSocket};
 
 type RustlsStream = rustls::StreamOwned<ServerSession, TcpStream>;
 type RustlsWebSocket = WebSocket<RustlsStream>;
 
-pub struct TungsteniteWsConnection {
-	web_socket: RustlsWebSocket,
+/// A web-socket connection object.
+pub struct TungsteniteWsConnection<Handler> {
+	tls_stream: Option<RustlsStream>,
+	web_socket: Option<RustlsWebSocket>,
+	connection_token: Token,
+	connection_handler: Arc<Handler>,
+	is_closed: bool,
```

Comment:
This struct is where a lot of the magic happens. The basis for the code was taken from the mio sgx example: https://github.com/apache/incubator-teaclave-sgx-sdk/blob/master/samplecode/mio/server/enclave/src/lib.rs

So quite a few code snippets and naming are taken from that example.

Regarding the two `Option<>` members:
When creating a new connection, it is just a regular TCP stream. We then need to do the TLS handshake on top of that. And as a last step, we do the web-socket handshake. In the first two stages, the `tls_stream` is `Some()` and `web_socket` is `None`. After the web-socket handshake, the `web_socket` is `Some()` and the `tls_stream` is `None`. This logic is captured in a convenience function `get_active_stream()`.

Reply:
I think this is fine for me, it would be more rust idiomatic though I guess if we have two connection type structs:

```
TungsteniteTlsConnection {
	tls_stream: Option<RustlsStream>,
	connection_token: Token,
	connection_handler: Arc<Handler>,
	is_closed: bool,
}
```

And this struct does then perform the websocket handshake and returns the 


```TungsteniteWsConnection {
tls_stream: Option<RustlsStream>,
	web_socket: Option<RustlsWebSocket>,
	connection_token: Token,
	connection_handler: Arc<Handler>,
	is_closed: bool,
}
```

This give better separation of concerns and make sure that we only pass the actual websocket connect into something that expects that the handshake is already performed. What do you think of that?

Reply:
The problem with this approach is, that the server manages the connections and stores them in a hash map - which takes one 'type' only. Furthermore, from the server's perspective, it does not know if a connection is an established TLS connection or an upgraded web-socket connection. That transition is made inside the connection itself.

Reply:
I totally agree that this is not an elegant solution, and it itches me too. But under the time restrictions, I could not find a better solution. Maybe when we come back to this some time in the future..